File: backend/handlers/notes.go
package handlers

import (
	"database/sql"
	"log"
	"basidian/database"
	"time"

	"github.com/gin-gonic/gin"
)

type Note struct {
	ID        string  `json:"id"`
	Title     string  `json:"title"`
	Content   string  `json:"content"`
	Date      string  `json:"date"`
	CreatedAt *string `json:"created_at"`
	UpdatedAt *string `json:"updated_at"`
}

type NoteRequest struct {
	Title   string `json:"title"`
	Content string `json:"content"`
	Date    string `json:"date"`
}

func GetNotes(c *gin.Context) {
	log.Printf("GetNotes: Fetching all notes")

	rows, err := database.DB.Query(`
		SELECT id, title, content, date, created, updated
		FROM notes
		ORDER BY date DESC
	`)
	if err != nil {
		log.Printf("GetNotes: Query failed: %v", err)
		c.JSON(500, gin.H{"error": "Failed to fetch notes"})
		return
	}
	defer rows.Close()

	notes := make([]Note, 0)
	for rows.Next() {
		var n Note
		var created, updated string
		if err := rows.Scan(&n.ID, &n.Title, &n.Content, &n.Date, &created, &updated); err != nil {
			log.Printf("GetNotes: Scan failed: %v", err)
			continue
		}
		if created != "" {
			n.CreatedAt = &created
		}
		if updated != "" {
			n.UpdatedAt = &updated
		}
		notes = append(notes, n)
	}

	log.Printf("GetNotes: Found %d notes", len(notes))
	c.JSON(200, notes)
}

func GetNote(c *gin.Context) {
	id := c.Param("id")

	var n Note
	var created, updated string
	err := database.DB.QueryRow(`
		SELECT id, title, content, date, created, updated
		FROM notes
		WHERE id = ?
	`, id).Scan(&n.ID, &n.Title, &n.Content, &n.Date, &created, &updated)

	if err == sql.ErrNoRows {
		c.JSON(404, gin.H{"error": "Note not found"})
		return
	}
	if err != nil {
		c.JSON(500, gin.H{"error": "Failed to fetch note"})
		return
	}

	if created != "" {
		n.CreatedAt = &created
	}
	if updated != "" {
		n.UpdatedAt = &updated
	}

	c.JSON(200, n)
}

func CreateNote(c *gin.Context) {
	var req NoteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": "Invalid JSON"})
		return
	}

	id := database.GenerateID()
	now := time.Now().Format(time.RFC3339)

	date := req.Date
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}

	_, err := database.DB.Exec(`
		INSERT INTO notes (id, title, content, date, created, updated)
		VALUES (?, ?, ?, ?, ?, ?)
	`, id, req.Title, req.Content, date, now, now)

	if err != nil {
		log.Printf("CreateNote: Insert failed: %v", err)
		c.JSON(500, gin.H{"error": "Failed to create note"})
		return
	}

	note := Note{
		ID:        id,
		Title:     req.Title,
		Content:   req.Content,
		Date:      date,
		CreatedAt: &now,
		UpdatedAt: &now,
	}

	log.Printf("CreateNote: Created note %s", id)
	c.JSON(201, note)
}

func UpdateNote(c *gin.Context) {
	id := c.Param("id")

	var req NoteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": "Invalid JSON"})
		return
	}

	// Check if note exists
	var exists bool
	err := database.DB.QueryRow("SELECT 1 FROM notes WHERE id = ?", id).Scan(&exists)
	if err == sql.ErrNoRows {
		c.JSON(404, gin.H{"error": "Note not found"})
		return
	}

	now := time.Now().Format(time.RFC3339)

	_, err = database.DB.Exec(`
		UPDATE notes
		SET title = ?, content = ?, date = ?, updated = ?
		WHERE id = ?
	`, req.Title, req.Content, req.Date, now, id)

	if err != nil {
		log.Printf("UpdateNote: Update failed: %v", err)
		c.JSON(500, gin.H{"error": "Failed to update note"})
		return
	}

	// Fetch updated record
	var n Note
	var created, updated string
	database.DB.QueryRow(`
		SELECT id, title, content, date, created, updated
		FROM notes WHERE id = ?
	`, id).Scan(&n.ID, &n.Title, &n.Content, &n.Date, &created, &updated)

	if created != "" {
		n.CreatedAt = &created
	}
	if updated != "" {
		n.UpdatedAt = &updated
	}

	log.Printf("UpdateNote: Updated note %s", id)
	c.JSON(200, n)
}

func DeleteNote(c *gin.Context) {
	id := c.Param("id")

	// Check if note exists
	var exists bool
	err := database.DB.QueryRow("SELECT 1 FROM notes WHERE id = ?", id).Scan(&exists)
	if err == sql.ErrNoRows {
		c.JSON(404, gin.H{"error": "Note not found"})
		return
	}

	_, err = database.DB.Exec("DELETE FROM notes WHERE id = ?", id)
	if err != nil {
		log.Printf("DeleteNote: Delete failed: %v", err)
		c.JSON(500, gin.H{"error": "Failed to delete note"})
		return
	}

	log.Printf("DeleteNote: Deleted note %s", id)
	c.Status(204)
}

func GetNotesByDate(c *gin.Context) {
	dateStr := c.Param("date")

	_, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		c.JSON(400, gin.H{"error": "Invalid date format. Use YYYY-MM-DD"})
		return
	}

	rows, err := database.DB.Query(`
		SELECT id, title, content, date, created, updated
		FROM notes
		WHERE date LIKE ?
		ORDER BY created DESC
	`, dateStr+"%")
	if err != nil {
		c.JSON(500, gin.H{"error": "Failed to fetch notes"})
		return
	}
	defer rows.Close()

	notes := make([]Note, 0)
	for rows.Next() {
		var n Note
		var created, updated string
		if err := rows.Scan(&n.ID, &n.Title, &n.Content, &n.Date, &created, &updated); err != nil {
			continue
		}
		if created != "" {
			n.CreatedAt = &created
		}
		if updated != "" {
			n.UpdatedAt = &updated
		}
		notes = append(notes, n)
	}

	c.JSON(200, notes)
}

func SearchNotes(c *gin.Context) {
	query := c.Query("q")
	if query == "" {
		c.JSON(400, gin.H{"error": "Search query is required"})
		return
	}

	searchPattern := "%" + query + "%"

	rows, err := database.DB.Query(`
		SELECT id, title, content, date, created, updated
		FROM notes
		WHERE title LIKE ? OR content LIKE ?
		ORDER BY date DESC
	`, searchPattern, searchPattern)
	if err != nil {
		c.JSON(500, gin.H{"error": "Failed to search notes"})
		return
	}
	defer rows.Close()

	notes := make([]Note, 0)
	for rows.Next() {
		var n Note
		var created, updated string
		if err := rows.Scan(&n.ID, &n.Title, &n.Content, &n.Date, &created, &updated); err != nil {
			continue
		}
		if created != "" {
			n.CreatedAt = &created
		}
		if updated != "" {
			n.UpdatedAt = &updated
		}
		notes = append(notes, n)
	}

	c.JSON(200, notes)
}

File: backend/src/basidian_server/handlers/notes.py
from datetime import datetime

from fastapi import APIRouter, HTTPException, Query
from loguru import logger

from .. import database as db
from ..models import Note, NoteRequest

router = APIRouter()


def _row_to_note(row) -> Note:
    """Convert a database row to a Note model."""
    created = row["created"] if row["created"] else None
    updated = row["updated"] if row["updated"] else None
    return Note(
        id=row["id"],
        title=row["title"],
        content=row["content"],
        date=row["date"],
        created_at=created,
        updated_at=updated,
    )


@router.get("/api/notes")
async def get_notes() -> list[Note]:
    """Fetch all notes ordered by date descending."""
    logger.info("GetNotes: Fetching all notes")

    assert db.db is not None
    async with db.db.execute("""
        SELECT id, title, content, date, created, updated
        FROM notes
        ORDER BY date DESC
    """) as cursor:
        rows = await cursor.fetchall()

    notes = [_row_to_note(row) for row in rows]
    logger.info(f"GetNotes: Found {len(notes)} notes")
    return notes


@router.post("/api/notes", status_code=201)
async def create_note(req: NoteRequest) -> Note:
    """Create a new note."""
    assert db.db is not None

    note_id = db.generate_id()
    now = datetime.now().isoformat()

    date = req.date if req.date else datetime.now().strftime("%Y-%m-%d")

    await db.db.execute(
        """
        INSERT INTO notes (id, title, content, date, created, updated)
        VALUES (?, ?, ?, ?, ?, ?)
        """,
        (note_id, req.title, req.content, date, now, now),
    )
    await db.db.commit()

    logger.info(f"CreateNote: Created note {note_id}")
    return Note(
        id=note_id,
        title=req.title,
        content=req.content,
        date=date,
        created_at=now,
        updated_at=now,
    )


@router.get("/api/notes/{note_id}")
async def get_note(note_id: str) -> Note:
    """Get a single note by ID."""
    assert db.db is not None

    async with db.db.execute(
        """
        SELECT id, title, content, date, created, updated
        FROM notes
        WHERE id = ?
        """,
        (note_id,),
    ) as cursor:
        row = await cursor.fetchone()

    if row is None:
        raise HTTPException(status_code=404, detail="Note not found")

    return _row_to_note(row)


@router.put("/api/notes/{note_id}")
async def update_note(note_id: str, req: NoteRequest) -> Note:
    """Update an existing note."""
    assert db.db is not None

    # Check if note exists
    async with db.db.execute(
        "SELECT 1 FROM notes WHERE id = ?", (note_id,)
    ) as cursor:
        if await cursor.fetchone() is None:
            raise HTTPException(status_code=404, detail="Note not found")

    now = datetime.now().isoformat()

    await db.db.execute(
        """
        UPDATE notes
        SET title = ?, content = ?, date = ?, updated = ?
        WHERE id = ?
        """,
        (req.title, req.content, req.date, now, note_id),
    )
    await db.db.commit()

    # Fetch updated record
    async with db.db.execute(
        """
        SELECT id, title, content, date, created, updated
        FROM notes WHERE id = ?
        """,
        (note_id,),
    ) as cursor:
        row = await cursor.fetchone()

    logger.info(f"UpdateNote: Updated note {note_id}")
    return _row_to_note(row)


@router.delete("/api/notes/{note_id}", status_code=204)
async def delete_note(note_id: str) -> None:
    """Delete a note."""
    assert db.db is not None

    # Check if note exists
    async with db.db.execute(
        "SELECT 1 FROM notes WHERE id = ?", (note_id,)
    ) as cursor:
        if await cursor.fetchone() is None:
            raise HTTPException(status_code=404, detail="Note not found")

    await db.db.execute("DELETE FROM notes WHERE id = ?", (note_id,))
    await db.db.commit()

    logger.info(f"DeleteNote: Deleted note {note_id}")


@router.get("/api/notes/date/{date}")
async def get_notes_by_date(date: str) -> list[Note]:
    """Get notes for a specific date."""
    # Validate date format
    try:
        datetime.strptime(date, "%Y-%m-%d")
    except ValueError:
        raise HTTPException(
            status_code=400, detail="Invalid date format. Use YYYY-MM-DD"
        )

    assert db.db is not None
    async with db.db.execute(
        """
        SELECT id, title, content, date, created, updated
        FROM notes
        WHERE date LIKE ?
        ORDER BY created DESC
        """,
        (f"{date}%",),
    ) as cursor:
        rows = await cursor.fetchall()

    return [_row_to_note(row) for row in rows]


@router.get("/api/search")
async def search_notes(q: str = Query(..., min_length=1)) -> list[Note]:
    """Search notes by title or content."""
    assert db.db is not None

    search_pattern = f"%{q}%"

    async with db.db.execute(
        """
        SELECT id, title, content, date, created, updated
        FROM notes
        WHERE title LIKE ? OR content LIKE ?
        ORDER BY date DESC
        """,
        (search_pattern, search_pattern),
    ) as cursor:
        rows = await cursor.fetchall()

    return [_row_to_note(row) for row in rows]
